tests/test_recursive.py

```diff
@@ -0,0 +1,253 @@
+from argparse import ArgumentError
+from dataclasses import dataclass, field
+
+from pytest import mark, raises, warns
+
+from pydargs import parse
+
+
+@dataclass
+class SubConfig:
+    a: int = 42
+    b: str = field(default="abc", metadata=dict(help="a string"))
+
+
+@dataclass
+class SubConfigWithRequiredField:
+    a: int
+    b: str = "abc"
+
+
+@dataclass
+class SubConfigWithPositional:
+    a: int = field(default=42, metadata=dict(positional=True))
+    b: str = "abc"
+
+
+@dataclass
+class NestedSub:
+    a: int = 42
+    b: SubConfig = field(default_factory=SubConfig)
+    c: str = "abc"
+
+
+class TestParseSubConfig:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfig = field(default_factory=SubConfig)
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    @mark.parametrize(
+        "help_string",
+        ["usage: prog [-h] [--sub-a SUB_A] [--sub-b SUB_B]", "sub:", "--sub-b SUB_B      a string (default: abc)"],
+    )
+    def test_help(self, capsys, help_string: str):
+        with raises(SystemExit):
+            parse(self.Config, ["--help"], prog="prog")  # type: ignore
+        captured = capsys.readouterr()
+        assert help_string in captured.out.replace("\n", "")
+
+    def test_parse(self, recwarn):
+        config = parse(self.Config, ["mode"])
+        assert len(recwarn) == 0
+        assert config.mode == "mode"
+        assert config.sub.a == 42
+        assert config.sub.b == "abc"
+        assert config.flag is False
+
+    def test_flag(self):
+        config = parse(self.Config, ["mode", "--flag"])
+        assert config.mode == "mode"
+        assert config.sub.a == 42
+        assert config.sub.b == "abc"
+        assert config.flag is True
+
+    def test_set_sub(self):
+        config = parse(self.Config, ["mode", "--sub-a", "1"])
+        assert config.mode == "mode"
+        assert config.sub.a == 1
+        assert config.sub.b == "abc"
+        assert config.flag is False
+
+
+class TestSubConfigCollision:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfig = field(default_factory=SubConfig)
+        sub_a: float = field(default=1.0)
+
+    def test_parse(self):
+        with raises(ArgumentError):
+            parse(self.Config, ["mode"])
```

Comment:
is this argparse raising an error when duplicate keys are added?

Reply:
Exactly! We need to verify that this already fails when creating the parser, and not (only) when we actually use the arguments.